File: src/schemas/analytics.py
from datetime import datetime
from typing import Optional, Any, Dict, List
from uuid import UUID
from pydantic import BaseModel, Field


class AnalyticsEventCreate(BaseModel):
    event_name: str = Field(..., max_length=255)
    event_type: str = Field(..., max_length=50)
    user_id: Optional[UUID] = None
    session_id: Optional[str] = None
    institution_id: Optional[UUID] = None
    properties: Optional[Dict[str, Any]] = None
    user_agent: Optional[str] = None
    ip_address: Optional[str] = None
    referrer: Optional[str] = None
    url: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None


class AnalyticsEventResponse(BaseModel):
    id: UUID
    event_name: str
    event_type: str
    user_id: Optional[UUID]
    session_id: Optional[str]
    properties: Optional[Dict[str, Any]]
    created_at: datetime

    class Config:
        from_attributes = True


class PerformanceMetricCreate(BaseModel):
    metric_name: str = Field(..., max_length=100)
    metric_value: float
    user_id: Optional[UUID] = None
    session_id: Optional[str] = None
    url: Optional[str] = None
    rating: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class PerformanceMetricResponse(BaseModel):
    id: UUID
    metric_name: str
    metric_value: float
    rating: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class UserSessionCreate(BaseModel):
    session_id: str
    user_id: Optional[UUID] = None
    institution_id: Optional[UUID] = None
    landing_page: Optional[str] = None
    referrer: Optional[str] = None
    device_type: Optional[str] = None
    browser: Optional[str] = None
    os: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None


class UserSessionUpdate(BaseModel):
    last_seen: Optional[datetime] = None
    page_views: Optional[int] = None
    events_count: Optional[int] = None
    exit_page: Optional[str] = None


class UserSessionResponse(BaseModel):
    id: UUID
    session_id: str
    user_id: Optional[UUID]
    first_seen: datetime
    last_seen: datetime
    page_views: int
    events_count: int
    landing_page: Optional[str]
    device_type: Optional[str]

    class Config:
        from_attributes = True


class FeatureUsageCreate(BaseModel):
    feature_name: str
    user_id: Optional[UUID] = None
    institution_id: Optional[UUID] = None
    properties: Optional[Dict[str, Any]] = None


class FeatureUsageResponse(BaseModel):
    id: UUID
    feature_name: str
    user_id: Optional[UUID]
    usage_count: int
    first_used_at: datetime
    last_used_at: datetime

    class Config:
        from_attributes = True


class BatchAnalyticsRequest(BaseModel):
    events: List[AnalyticsEventCreate] = Field(default_factory=list)
    performance_metrics: List[PerformanceMetricCreate] = Field(default_factory=list)


class AnalyticsDashboardStats(BaseModel):
    total_users: int
    active_users_today: int
    active_users_week: int
    active_users_month: int
    total_sessions: int
    avg_session_duration: float
    total_page_views: int
    avg_pages_per_session: float


class FeatureAdoptionStats(BaseModel):
    feature_name: str
    total_users: int
    total_usage: int
    unique_users_today: int
    unique_users_week: int
    unique_users_month: int
    adoption_rate: float


class UserFlowNode(BaseModel):
    page: str
    count: int
    drop_off_rate: float


class UserFlowAnalysis(BaseModel):
    nodes: List[UserFlowNode]
    total_sessions: int


class RetentionCohort(BaseModel):
    cohort_date: str
    users_count: int
    retention_day_1: float
    retention_day_7: float
    retention_day_14: float
    retention_day_30: float


class TopEventStats(BaseModel):
    event_name: str
    event_type: str
    count: int
    unique_users: int


class PerformanceStats(BaseModel):
    metric_name: str
    avg_value: float
    p50_value: float
    p75_value: float
    p95_value: float
    good_count: int
    needs_improvement_count: int
    poor_count: int

File: frontend/src/api/analytics.ts
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_BASE_URL || '';

export interface AnalyticsDashboardStats {
  total_users: number;
  active_users_today: number;
  active_users_week: number;
  active_users_month: number;
  total_sessions: number;
  avg_session_duration: number;
  total_page_views: number;
  avg_pages_per_session: number;
}

export interface FeatureAdoptionStats {
  feature_name: string;
  total_users: number;
  total_usage: number;
  unique_users_today: number;
  unique_users_week: number;
  unique_users_month: number;
  adoption_rate: number;
}

export interface UserFlowNode {
  page: string;
  count: number;
  drop_off_rate: number;
}

export interface UserFlowAnalysis {
  nodes: UserFlowNode[];
  total_sessions: number;
}

export interface RetentionCohort {
  cohort_date: string;
  users_count: number;
  retention_day_1: number;
  retention_day_7: number;
  retention_day_14: number;
  retention_day_30: number;
}

export interface TopEventStats {
  event_name: string;
  event_type: string;
  count: number;
  unique_users: number;
}

export interface PerformanceStats {
  metric_name: string;
  avg_value: number;
  p50_value: number;
  p75_value: number;
  p95_value: number;
  good_count: number;
  needs_improvement_count: number;
  poor_count: number;
}

export const analyticsApi = {
  getDashboardStats: async (institutionId?: string): Promise<AnalyticsDashboardStats> => {
    const params = institutionId ? { institution_id: institutionId } : {};
    const response = await axios.get(`${API_URL}/api/v1/analytics/dashboard`, { params });
    return response.data;
  },

  getFeatureAdoption: async (
    institutionId?: string,
    limit = 20
  ): Promise<FeatureAdoptionStats[]> => {
    const params = { limit, ...(institutionId ? { institution_id: institutionId } : {}) };
    const response = await axios.get(`${API_URL}/api/v1/analytics/features/adoption`, {
      params,
    });
    return response.data;
  },

  getUserFlow: async (institutionId?: string, limit = 10): Promise<UserFlowAnalysis> => {
    const params = { limit, ...(institutionId ? { institution_id: institutionId } : {}) };
    const response = await axios.get(`${API_URL}/api/v1/analytics/user-flow`, { params });
    return response.data;
  },

  getRetentionCohorts: async (
    institutionId?: string,
    cohortDays = 30
  ): Promise<RetentionCohort[]> => {
    const params = {
      cohort_days: cohortDays,
      ...(institutionId ? { institution_id: institutionId } : {}),
    };
    const response = await axios.get(`${API_URL}/api/v1/analytics/retention/cohorts`, {
      params,
    });
    return response.data;
  },

  getTopEvents: async (institutionId?: string, limit = 20): Promise<TopEventStats[]> => {
    const params = { limit, ...(institutionId ? { institution_id: institutionId } : {}) };
    const response = await axios.get(`${API_URL}/api/v1/analytics/events/top`, { params });
    return response.data;
  },

  getPerformanceStats: async (metricName?: string, days = 7): Promise<PerformanceStats[]> => {
    const params = { days, ...(metricName ? { metric_name: metricName } : {}) };
    const response = await axios.get(`${API_URL}/api/v1/analytics/performance/stats`, {
      params,
    });
    return response.data;
  },
};
